### PR TITLE
New script - PcapFileExtractor

### PR DESCRIPTION
**Pull request by Shellyber**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/25215

## Description
This automation extracts all possible files from a PCAP file.

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [x] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    elif protocol == 'imf':
+        command_ = f'tshark -r {file_path} --export-objects imf,{dir_path}'
+
+    elif protocol == 'tftp':
+        command_ = f'tshark -r {file_path} --export-objects tftp,{dir_path}'
+
+    elif protocol == 'dicom':
+        command_ = f'tshark -r {file_path} --export-objects dicom,{dir_path}'
+
+    process = subprocess.Popen(command_, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:94]: do not use `shell=True`

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        return_error('You must set entry_id')
+
+    res = demisto.executeCommand('getFilePath', {'id': file_entry_id})
+    if res[0]['Type'] == entryTypes['error']:
+        return_error(f'Failed to get the file path for entry: {file_entry_id}')
+
+    return res[0]['Contents']['path']
+
+
+def find_files_protocol(file_path):
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:32]: i dont think you need that function, you can just run `--export-objects` on all protcols

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    elif protocol == 'tftp':
+        command_ = f'tshark -r {file_path} --export-objects tftp,{dir_path}'
+
+    elif protocol == 'dicom':
+        command_ = f'tshark -r {file_path} --export-objects dicom,{dir_path}'
+
+    process = subprocess.Popen(command_, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
+    process.communicate()
+    excluded_files = [f for f in glob.glob(dir_path + '/*')]
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:96]: why is this needed?

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+def upload_files(excluded_files, dir_path, file_path):
+    """
+
+    :param excluded_files: excluded files
+    :param dir_path: dir path for the files
+    :param file_path: the path to the pcap file
+    :return:
+    """
+    filenames = []  # type: ignore
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:109]: why the `ignore` here?

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    else:
+        results = []
+        context = []
+        protocol, packet_data = find_files_protocol(file_path)
+
+        md5 = hashlib.md5()
+        sha1 = hashlib.sha1()
+        sha256 = hashlib.sha256()
+        files_base_names = [os.path.basename(file_path) for file_path in filenames]  # noqa[F812]
+        files_dic = {file_path: os.path.basename(file_path) for file_path in filenames}
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:130]: not sure why we're iterating the same object 3 times, twice here and 1 in the loop below
lets just iterate it once

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+        for file_path, file_name in files_dic.items():
+            for data in packet_data:
+                packet_number = data.split()[0]
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:134]: `packet_number` is defined here and in the loop below, why?

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                for packet_number in packet_data:
+                    data = [i for i in packet_number.split()]
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:136]:
```suggestion
                    data = packet_number.split()
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                    source_ip = data[2]
+                    dest_ip = data[4]
+
+            with open(file_path, 'rb') as _file:
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:140]:
```suggestion
            with open(file_path, 'rb') as file:
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                demisto.results(fileResult(file_name, _file.read()))
+
+            with open(file_path, 'rb') as _file:
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:143]: why are you opening the file twice?

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                'FileExtension': os.path.splitext(file_name)[1],
+                'SourceIP': source_ip,
+                'DestinationIP': dest_ip
+            })
+
+            ec = {
+                'PcapExtractedFiles(val.FileMD5 === obj.FileMD5)': context
+            }
+
+        results.append(
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:165]: lets use `CommandResults`

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                'Type': entryTypes['note'],
+                'ContentsFormat': formats['json'],
+                'Contents': {'extractedFiles': files_base_names},
+                'EntryContext': ec,
+                'ReadableContentsFormat': formats['markdown'],
+                'HumanReadable': tableToMarkdown('Pcap Extracted Files',
+                                                 [{'name': file_name} for file_name in files_base_names])
+            })
+
+        demisto.results(results)
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:176]: lets remove this line and call `return_outputs` in the `main` func so it will be easier for unti testing

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+            ec = {
+                'PcapExtractedFiles(val.FileMD5 === obj.FileMD5)': context
+            }
+
+        results.append(
+            {
+                'Type': entryTypes['note'],
+                'ContentsFormat': formats['json'],
+                'Contents': {'extractedFiles': files_base_names},
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:169]:
```suggestion
                'Contents': ec,
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        args = demisto.args()
+        file_path = get_pcap_path(args)
+        find_files_protocol(file_path)
+        excluded_files = extract_files(file_path=file_path, dir_path=dir_path)
+        upload_files(excluded_files, dir_path, file_path)
+
+    except Exception as e:
+        return_error(f'Failed to execute PcapFileExtracor. Error: {str(e)}')
+    finally:
+        shutil.rmtree(dir_path)
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:191]: check if `dir_path` before to make sure it exists

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+
+
+if __name__ in ('__builtin__', 'builtins'):
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:194]:
```suggestion
if __name__ in ('__builtin__', 'builtins', '__main__'):
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.yml [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+- contextPath: PcapExtractedFiles.FileName
+  description: File name.
+  type: String
+- contextPath: PcapExtractedFiles.FileSize
+  description: File size.
+  type: Number
+- contextPath: PcapExtractedFiles.FileType
+  description: File type.
+  type: String
+- contextPath: PcapExtractedFiles.EntryID
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.yml:24]: remove this

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.yml [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+- contextPath: PcapExtractedFiles.FileMD5
+  description: The MD5 hash of the file.
+  type: String
+- contextPath: PcapExtractedFiles.FileSHA1
+  description: The SHA1 hash of the file.
+  type: String
+- contextPath: PcapExtractedFiles.FileSHA256
+  description: The SHA256 hash of the file.
+  type: String
+- contextPath: PcapExtractedFiles.FileExtensaion
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.yml:45]:
```suggestion
- contextPath: PcapExtractedFiles.FileExtension
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_description.md [hunk 17]
```diff
@@ -0,0 +1,2 @@
+## PcapFileExtractor Script
+This automation extracts all possible files from a PCAP file.
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_description.md:2]: no need for description files for scripts

File: Pipfile [hunk 18]
```diff
@@ -23,6 +23,7 @@ beautifulsoup4 = "==4.8.1"
 pypdf2 = "==1.26.0"
 freezegun = "==0.3.12"
 vulture = "==1.5"
+pyshark = "*"
```
Comment (Itay4) [inline Pipfile:26]: dont add this here, add a pipfile and a pipfile.lock to the script dir

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
+import demistomock as demisto
+from CommonServerPython import *  # noqa: E402 lgtm [py/polluting-import]
+from CommonServerUserPython import *  # noqa: E402 lgtm [py/polluting-import]
+import os
+from os.path import isfile
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:5]: `os` is already imported in commonserver, you can remove

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py (hunk 9 above)
Comment (Shellyber) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:143]: Tried to do it in the same opening of the file, but it doesn't work. 
Seems to lose data in the process

Comment (Itay4) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:143]: you cant call `read()` twice, call it once and store in variable to use

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+
+        demisto.results(results)
+
+
+def main():
+    dir_path = mkdtemp()
+    try:
+        args = demisto.args()
+        file_path = get_pcap_path(args)
+        find_files_protocol(file_path)
```
[the remark this answers is not in the document]
Comment (Shellyber) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:184]: No reason, forgot to remove.

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py (hunk 2 above)
Comment (Shellyber) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:32]: I can't, it doesn't work for all protocols in the command.
You must specify one protocol and it should be the right as the type of the file.

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py (hunk 1 above)
Comment (Shellyber) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:94]: Fixed it using shlex

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py (hunk 3 above)
Comment (Shellyber) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:96]: This is the list of all files exported from the pcap. we iterate over them in the upload command.

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_test.py [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+from PcapFileExtractor import find_files_protocol, extract_files
+
+
+def test_find_protocol():
+    file_path = './TestData/tftp_rrq.pcap'
+    protocol, _ = find_files_protocol(file_path)
+    assert 'tftp' == protocol
+
+
+def test_extract_files():
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_test.py:13]: lets use `tmpdir`

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.yml [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+script: '-'
+subtype: python3
+system: false
+tags:
+- pcap
+- file
+- Utility
+timeout: '0'
+type: python
+dockerimage: demisto/pcap-miner:1.0.0.9360
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.yml:54]: add fromversion 5.0

**Comment by kirbles19** [comment]
@Shellyber Reviewed the YML, didn't see a releasenotes file.

**Comment by Shellyber** [comment]
@kirbles19 Thanks Alex, no need for release notes since it's new script

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+                try:
+                    source_ip = packet[2]
+                    dest_ip = packet[4]
+                except Exception:
+                    pass
+
+            with open(file_path, 'rb') as file:
+                demisto.results(fileResult(file_name, file.read()))
+
+                data = file.read()
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:113]: no need to `read` twice
initialize `data` before the call to `demisto.results` and use it

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+                'FileExtension': os.path.splitext(file_name)[1],
+                'SourceIP': source_ip,
+                'DestinationIP': dest_ip
+            })
+
+        readable_output = tableToMarkdown('Pcap Extracted Files', [{'name': file_name} for file_name in files])
+
+        results = CommandResults(
+            outputs_prefix='PcapExtractedFiles',
+            outputs_key_field='FileName',
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:134]: better to use one of the hashes, file name is not unique

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_test.py [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+from PcapFileExtractor import find_files_packets, upload_files
+
+
+def test_find_protocol():
+    file_path = './TestData/tftp_rrq.pcap'
+    protocol, _ = find_files_packets(file_path)
+    assert 'tftp' == protocol
+
+
+def test_extract_files(tmpdir):
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_test.py:12]: add test docstring

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_test.py [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+    file_path = 'TestData/tftp_rrq.pcap'
+    results = upload_files(file_path=file_path, dir_path=tmpdir)
+    assert 'Pcap Extracted Files' in results.readable_output
+    assert os.path.isfile(os.path.join(tmpdir, 'rfc1350.txt'))
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor_test.py:16]: validate also context outputs please

File: Pipfile (hunk 18 above)
Comment (Itay4) [reply Pipfile:26]: revert this

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+        return_error('You must set entry_id')
+
+    res = demisto.executeCommand('getFilePath', {'id': file_entry_id})
+    if isError(res):
+        raise Exception(f'Failed to get the file path for entry: {file_entry_id}')
+
+    return res[0]['Contents']['path']
+
+
+def find_files_packets(file_path):
```
[the remark this answers is not in the document]
Comment (Shellyber) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:30]: removing this command entirely

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/README.md [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+## Outputs
+---
+
+| **Path** | **Description** | **Type** |
+| --- | --- | --- |
+| PcapExtractedFiles.FileName | File name. | String |
+| PcapExtractedFiles.FileSize | File size. | Number |
+| PcapExtractedFiles.FileType | File type. | String |
+| PcapExtractedFiles.EntryID | The entry ID of the file. | String |
+| PcapExtractedFiles.SourceIP | Extracted file source IP. | String |
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/README.md:27]: update outputs please

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+    process.communicate()
+
+    context = []
+
+    md5 = hashlib.md5()
+    sha1 = hashlib.sha1()
+    sha256 = hashlib.sha256()
+    for root, _, files in os.walk(dir_path):
+        if len(files) == 0:
+            return_error('Could not find files')
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:54]: no need to return error, just regular entry

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+    process.communicate()
+
+    context = []
+
+    md5 = hashlib.md5()
+    sha1 = hashlib.sha1()
+    sha256 = hashlib.sha256()
+    for root, _, files in os.walk(dir_path):
+        if len(files) == 0:
+            return 'could not find files'
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:54]:
```suggestion
            return 'No files found.'
```
